storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -122,31 +139,31 @@ public void close() throws IOException {
         closed = true;
     }
 
-    private void flushBuffer(final int offset,
-                             final int actualPartSize) throws IOException {
+    private void flushBuffer(final ByteBuffer partBuffer, final int actualPartSize) throws IOException {
         try {
-            final ByteArrayInputStream in = new ByteArrayInputStream(partBuffer.array(), offset, actualPartSize);
-            uploadPart(in, actualPartSize);
-            partBuffer.clear();
+            final byte[] array = new byte[actualPartSize];
+            partBuffer.get(array, 0, actualPartSize);
+
+            final var uploadPartRequest = new UploadPartRequest()
+                .withBucketName(bucketName)
+                .withKey(key)
+                .withUploadId(uploadId)
+                .withPartSize(actualPartSize)
+                .withPartNumber(partNumber.incrementAndGet())
+                .withInputStream(new ByteArrayInputStream(array));
+
+            // Run request async
+            partUploads = partUploads.thenCombine(
+                CompletableFuture.supplyAsync(() -> client.uploadPart(uploadPartRequest)),
+                (partETags, result) -> {
+                    partETags.add(result.getPartETag());
+                    return partETags;
+                });
         } catch (final Exception e) {
```

Comment:
RuntimeException maybe?

Reply:
Or do we need this catch here at all? At least if I remove the catch tests are passing and the exception I believe will be thrown on call to `get`

Reply:
I'm failing to see how this would help giving that RuntimeException is a subclass of Exception; could you elaborate?

Reply:
Agree. And we the fixes on #282 the transaction should be aborted properly.